app/ide-desktop/lib/profiling/src/profiling.ts

```diff
@@ -0,0 +1,386 @@
+}
+
+type Metadata = {
+    source: SourceLocation
+    profiling_level: string
+    label: string
+}
+
+/**
+ * Create a Metadata object that gets populated with the `SourceLocation` of the caller of the
+ * caller of this methods.
+ */
+function makeMetadata(profiling_level: string, label: string): Metadata {
+    const source = callerInfo(2)
+    return {
+        source,
+        profiling_level,
+        label,
+    }
+}
+
+/**
+ * Return the `SourceLocation` of this method call. By supplying the `caller_level` argument, the
+ * `SourceLocation` will be shifted up the call stack by that number of levels.
+ *
+ */
+function callerInfo(caller_level = 0): SourceLocation | null {
+    const e = new Error()
+    const regex = /\((.*):(\d+):(\d+)\)$/
+    const caller = e.stack.split('\n')[2 + caller_level]
+    if (caller === undefined) {
+        return null
+    }
+    const match = regex.exec(caller)
+    if (match === null) {
+        return null
+    }
+    assert(match.length >= 3, 'Regex was expected to produce three matches.')
+    return {
+        file: match[1],
+        line: Number(match[2]),
+    }
+}
+
+// ==============
+// === Labels ===
+// ==============
+
+/// Return a string that encodes the given log level and name for a mark that indicates the start of
+/// an interval. The label encoded the log level, name and event type.
+function startIntervalLabel(metadata: Metadata): string {
+    return `${metadata.label} [START]`
+}
+
+/// Return a string that encodes the given log level and name for a mark that indicates the end of
+/// an interval.
+function endIntervalLabel(metadata: Metadata): string {
+    return `${metadata.label} [END]`
+}
+
+/// Return a string that encodes the given log level and name for display in the developer console
+// performance tool..
+function performanceToolLabel(metadata: Metadata): string {
+    return `${metadata.label} (${metadata.source.file}:${metadata.source.line})`
+}
+
+// ================
+// === Profiler ===
+// ================
+
+/**
+ * Class that exposes `start`, `end` and `measure` methods, as well as some utility methods for
+ * working with the profiler.
+ */
+class Profiler {
+    logLevel: string
+    toggledInEnvironment: boolean
+
+    constructor(logLevel: string) {
+        this.logLevel = logLevel
+        // The value `process.env.PROFILING_LEVEL` is populated through the [environment plugin](https://webpack.js.org/plugins/environment-plugin/)
+        // at build time from an environment variable.
+        this.toggledInEnvironment = process.env.PROFILING_LEVEL.toLowerCase().includes(
+            logLevel.toLowerCase()
+        )
+
+        profilerRegistry.push(this)
+    }
+
+    logLevelName(): string {
+        return this.logLevel.toUpperCase()
+    }
+
+    /// Indicate whether this profiler should perform profiling.
+    isActive(): boolean {
+        // We need to check whether this logger, or a logger that is higher in the hierarchy, is activated.
```

Comment:
these are not loggers, these are profilers. Please check other docs for naming too.